os/kernel/task/task_exit.c

```diff
@@ -108,11 +108,15 @@ static void prepare_exit(struct tcb_s *tcb)
 #ifdef CONFIG_SMP
 	/* Make sure that the system knows about the locked state */
 
-	spin_setbit(&g_cpu_lockset, this_cpu(), &g_cpu_locksetlock, \
+	cpu_set_t lock = g_cpu_lockset;
+	lock &= (1 << this_cpu());
+	if (!lock) {
```

Comment:
There will be 2 condition able to enter this condition,
1) No CPU hold the locks
2) Another CPU hold the locks
If we are in second condition, is it okay to just invoke spin_setbit directly?

Reply:
I think some small changes can be made...?

`#ifdef CONFIG_SMP
	/* Make sure that the system knows about the locked state */
	cpu_set_t lock = g_cpu_lockset;
	lock &= (1 << this_cpu());
	if (!lock) {
		spin_setbit(&g_cpu_lockset, this_cpu(), &g_cpu_locksetlock, \
				&g_cpu_schedlock);
		lock = 1;
	}
#endif

	exit_tcb->lockcount++;

	mm_is_sem_available(exit_tcb);

	exit_tcb->lockcount--;
#ifdef CONFIG_SMP
	/* Make sure that the system knows about the locked state */
	if (exit_tcb->lockcount <= 0 && lock == 1) {
		spin_clrbit(&g_cpu_lockset, this_cpu(), &g_cpu_locksetlock, \
				&g_cpu_schedlock);
	}
#endif
}`

Reply:
Yes. You are right. I changed the whole logic to take the lock based on the current tcb lockcount.

Reply:
Looks good to me